### PR TITLE
Added changes to schbench stats in LKP

### PR DESCRIPTION
**Pull request by SSapkal**
Added changes to schbench stats to get the max and relative standard deviation of the end latency's between the iterations. Updated the spec/stats for the same.

Signed-off-by: Swapnil Sapkal <swapnil.sapkal@amd.com>

### CODE REVIEW

**Comment by SSapkal** [comment]
The ncompare results with the updated stats.
```
# lkp ncompare -s testcase=schbench/commit=6.0.7 -s testcase=schbench/commit=6.1.0-rc4 -a -k

=========================================================================================
compiler/iterations/kconfig/message_threads/rootfs/runtime/testcase/worker_threads:
  gcc/10x/defconfig/1/rhel/30s/schbench/32

commit:
  6.0.7
  6.1.0-rc4

           6.0.7                   6.1.0-rc4
---------------- ---------------------------
         %stddev     %change         %stddev
             \          |                \
     45.61            +4.3%      47.56 ±  3%  schbench.latency_50%_us
      5.56 ± 10%     -15.2%       4.72 ± 40%  schbench.latency_50%_us_stddev
     54.33            +2.5%      55.66 ±  2%  schbench.latency_75%_us
      2.42 ± 13%     +41.9%       3.44 ± 26%  schbench.latency_75%_us_stddev
     61.72            +1.4%      62.61        schbench.latency_90%_us
      1.86 ± 41%     +48.1%       2.76 ±  4%  schbench.latency_90%_us_stddev
    117.44 ± 25%      -6.2%     110.16 ±  9%  schbench.latency_95%_us
     37.05 ± 26%     +15.9%      42.94 ±  8%  schbench.latency_95%_us_stddev
    186.11 ± 10%      -1.9%     182.50 ±  2%  schbench.latency_99%_us
     15.90 ± 22%     +23.3%      19.60 ± 32%  schbench.latency_99%_us_stddev
    222.06           -10.6%     198.44        schbench.latency_99.5%_us
     14.24 ± 17%     +20.8%      17.20 ±  8%  schbench.latency_99.5%_us_stddev
    249.78 ±  5%     -14.4%     213.83        schbench.latency_99.9%_us
     14.64 ± 24%     +34.7%      19.71        schbench.latency_99.9%_us_stddev
    252.22 ±  5%      -4.0%     242.06        schbench.latency_max_us
     15.04 ± 21%     +44.2%      21.70 ± 17%  schbench.latency_max_us_stddev
     23.11 ±  3%      +2.9%      23.78        schbench.latency_min_us
     20.25 ± 61%     -81.2%       3.80 ± 19%  schbench.latency_min_us_stddev
    311.00            +3.9%     323.00        schbench.end_latency_99%_us_max
      7.70            -9.9%       6.94        schbench.end_latency_99%_us_stddev
```

File: stats/schbench [hunk 1]
```diff
@@ -33,3 +38,6 @@ stats.each do |k, v|
   puts "#{k}: #{v.average.round(2)}"
   puts "#{k}_stddev: #{v.relative_stddev.round(2)}"
 end
+
+puts "end_latency_99%_us_max: #{end_latency.max}"
```
Comment (rli9) [inline stats/schbench:42]: thanks, consult what the meaning of "end" is in end_latency?

**Review by yhuang-intel — COMMENTED** [review]
Why do we need last 99% latency?  Why average isn't good enough? And IIUC, you output max 99% latency instead of last.  You may need to output stats['latency_99%_us].last instead?

**Comment by yu-chen-surf** [comment]
> Hi @yhuang-intel,
> 
> > Why do we need last 99% latency?
> > As per our observation schbench threads are not properly collocated for the first 10-20 seconds. So the 10/20 secs 99% le latency value won't be the correct metric to look at. So we prefer to look at 99th le reported at end of 30th second.
> 
IMO it is reasonable to pick the last reported tail latency in every iteration.  I'm wondering why it took 10 - 20 seconds to warm up the system?  In theory the fork load balance should spread the tasks quite fast, and the cache warm up should also be very fast.  10 - 20 seconds seems to be abnormal. May I know what parameters of schbench you were running, and what the CPU topology of your system? Is it a server or a desktop? What is the boot message when boot with sched_verbose in the boot command line?

**Comment by yhuang-intel** [comment]
> > Hi @yhuang-intel,
> > > Why do we need last 99% latency?
> > > As per our observation schbench threads are not properly collocated for the first 10-20 seconds. So the 10/20 secs 99% le latency value won't be the correct metric to look at. So we prefer to look at 99th le reported at end of 30th second.
> 
> IMO it is reasonable to pick the last reported tail latency in every iteration. I'm wondering why it took 10 - 20 seconds to warm up the system? In theory the fork load balance should spread the tasks quite fast, and the cache warm up should also be very fast. 10 - 20 seconds seems to be abnormal. May I know what parameters of schbench you were running, and what the CPU topology of your system? Is it a server or a desktop? What is the boot message when boot with sched_verbose in the boot command line?

If so, I think we should change all statistics for all percentiles, not only 99%, but also 50%, 70%, ...

**Comment by SSapkal** [comment]
> > Hi @yhuang-intel,
> > > Why do we need last 99% latency?
> > > As per our observation schbench threads are not properly collocated for the first 10-20 seconds. So the 10/20 secs 99% le latency value won't be the correct metric to look at. So we prefer to look at 99th le reported at end of 30th second.
> 
> IMO it is reasonable to pick the last reported tail latency in every iteration. I'm wondering why it took 10 - 20 seconds to warm up the system? In theory the fork load balance should spread the tasks quite fast, and the cache warm up should also be very fast. 10 - 20 seconds seems to be abnormal. May I know what parameters of schbench you were running, and what the CPU topology of your system? Is it a server or a desktop? What is the boot message when boot with sched_verbose in the boot command line?

Hi @yu-chen-surf, On EPYC servers with Split-LLC, for lower counts of workers per messenger, the tail latency is determined by how quickly the messengers and their corresponding workers get co-located on to the CPUs of the same LLC domain (MC). Based on our analysis, this colocation rarely happens during task-wakeup, but instead happens during new-idle load-balancing under specific conditions.  Further, once the colocation happens we have seen instances of subsequent spread occuring which causes an increase in the tail latency. Hence we want to consider the 99th percentiles in all of 10s, 20s, 30s. We are ok taking the maximum value among these three.

**Comment by yu-chen-surf** [comment]
> > > Hi @yhuang-intel,
> > > > Why do we need last 99% latency?
> > > > As per our observation schbench threads are not properly collocated for the first 10-20 seconds. So the 10/20 secs 99% le latency value won't be the correct metric to look at. So we prefer to look at 99th le reported at end of 30th second.
> > 
> > 
> > IMO it is reasonable to pick the last reported tail latency in every iteration. I'm wondering why it took 10 - 20 seconds to warm up the system? In theory the fork load balance should spread the tasks quite fast, and the cache warm up should also be very fast. 10 - 20 seconds seems to be abnormal. May I know what parameters of schbench you were running, and what the CPU topology of your system? Is it a server or a desktop? What is the boot message when boot with sched_verbose in the boot command line?
> 
> Hi @yu-chen-surf, On EPYC servers with Split-LLC, for lower counts of workers per messenger, the tail latency is determined by how quickly the messengers and their corresponding workers get co-located on to the CPUs of the same LLC domain (MC). Based on our analysis, this colocation rarely happens during task-wakeup, but instead happens during new-idle load-balancing under specific conditions. Further, once the colocation happens we have seen instances of subsequent spread occuring which causes an increase in the tail latency. Hence we want to consider the 99th percentiles in all of 10s, 20s, 30s. We are ok taking the maximum value among these three.
Task wake up tries to find an available CPU within current LLC domain, even if it is split-LLC. If task wake up failed to spread the workers, it might be that WSYNC has taken effect, and the waker tries to put the wakee on current CPU without spreading them. However for a schbench wake up, it is based on futex, which does not have WSYNC flag set during wake up.  I suspect there might be a bug that fail to spread the workers. Anyway, this patch looks reasonable to me, but as Ying suggested, can 50%, 70%... also be adjusted as 99%?

**Comment by SSapkal** [comment]
> > > > Hi @yhuang-intel,
> > > > > Why do we need last 99% latency?
> > > > > As per our observation schbench threads are not properly collocated for the first 10-20 seconds. So the 10/20 secs 99% le latency value won't be the correct metric to look at. So we prefer to look at 99th le reported at end of 30th second.
> > > 
> > > 
> > > IMO it is reasonable to pick the last reported tail latency in every iteration. I'm wondering why it took 10 - 20 seconds to
> > > warm up the system? In theory the fork load balance should spread the tasks quite fast, and the cache warm up should also
> > > be very fast. 10 - 20 seconds seems to be abnormal. May I know what parameters of schbench you were running, and what
> > > the CPU topology of your system? Is it a server or a desktop? What is the boot message when boot with sched_verbose in
> > > the boot command line?
> > 
> > 
> > Hi @yu-chen-surf, On EPYC servers with Split-LLC, for lower counts of workers per messenger, the tail latency is determined > by how quickly the messengers and their corresponding workers get co-located on to the CPUs of the same LLC domain
> > (MC). Based on our analysis, this colocation rarely happens during task-wakeup, but instead happens during new-idle load
> > balancing under specific conditions. Further, once the colocation happens we have seen instances of subsequent spread
> > occuring which causes an increase in the tail latency. Hence we want to consider the 99th percentiles in all of 10s, 20s, 30s.
> > We are ok taking the maximum value among these three.
> 
> Task wake up tries to find an available CPU within current LLC domain, even if it is split-LLC. If task wake up failed to spread the workers, it might be that WSYNC has taken effect, and the waker tries to put the wakee on current CPU without spreading them. However for a schbench wake up, it is based on futex, which does not have WSYNC flag set during wake up. I suspect there might be a bug that fail to spread the workers. Can you give some the instruction you run schbench? Anyway, this patch looks reasonable to me, but as Ying suggested, can 50%, 70%... also be adjusted as 99%?

Initial task placement is spreading the schbench tasks correctly but workers are co-located because of the idle load balancing due to noise in the system such as a pinned kworker running on the same cpu as schbench worker. This is causing run to run variance in tail latencies.  This problem was described in LPC talk this year. 
[https://www.youtube.com/watch?v=M8MLLsKgQYg&t=7509s](url)
Sure, I will update the patch to get 50%, 75%, .. statistics.

**Comment by rli9** [comment]
thanks for the patch

**Pull request merged**